### PR TITLE
report: remove Util.UIStrings mutation, add I18n renderer class

### PR DESCRIPTION
**Pull request by connorjclark**
Cleanup for https://github.com/GoogleChrome/lighthouse/pull/9166/files#r361547882

### CODE REVIEW

File: lighthouse-core/report/html/renderer/report-renderer.js [hunk 1]
```diff
@@ -249,9 +243,6 @@ class ReportRenderer {
   }
 }
 
-/** @type {LH.I18NRendererStrings} */
-ReportRenderer._UIStringsStash = {};
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/report-renderer.js]: hmm

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 2]
```diff
@@ -25,14 +25,19 @@ const URL_PREFIXES = ['http://', 'https://', 'data:'];
 class DetailsRenderer {
   /**
    * @param {DOM} dom
+   * @param {LH.ReportResult} report
    */
-  constructor(dom) {
-    /** @type {DOM} */
+  constructor(dom, report) {
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/details-renderer.js]: easiest way to get the report strings everywhere

**Comment by connorjclark** [comment]
> and perhaps use the formatting from

does this include methods like `getRootDomain`?

**Comment by connorjclark** [comment]
>  perhaps use the formatting from

would that include getEmulationDescriptions?

**Comment by connorjclark** [comment]
alright, with this second pass I moved all the stuff that is mutable (strings, locale, formatters) to an `I18n` class, and made it globally accessible via `Util.i18n. This gets rid of the nasty mutations of `Util.UIStrings` without passing an extra object everywhere. Also, the audits that used some formatting now actually use the correct locale (before, it was just doing `en`).

File: lighthouse-core/lib/i18n/i18n.js [hunk 3]
```diff
@@ -337,14 +337,15 @@ function getRendererFormattedStrings(locale) {
   if (!localeMessages) throw new Error(`Unsupported locale '${locale}'`);
 
   const icuMessageIds = Object.keys(localeMessages).filter(f => f.includes('core/report/html/'));
-  /** @type {LH.I18NRendererStrings} */
+  /** @type {Record<string, string>} */
```
Comment (patrickhulce) [inline lighthouse-core/lib/i18n/i18n.js]: why this change? an explanation on the tsignore would be nice :)

File: lighthouse-core/report/html/renderer/i18n.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+/* globals self, URL */
+
+// Not named `NBSP` because that creates a duplicate identifier (util.js).
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/i18n.js]: 😬, on of these days the grimaces will add up to a build step for renderer :)

File: lighthouse-core/report/html/renderer/i18n.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  constructor(locale) {
+    this.setNumberDateLocale(locale);
+    this._numberDateLocale = locale;
+    this._numberFormatter = new Intl.NumberFormat(locale);
+  }
+
+  /**
+   * @param {LH.I18NRendererStrings} strings
+   */
+  setStrings(strings) {
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/i18n.js]: feels a little weird to pair this with a getter property, I would just expect `set strings()` or `getStrings()`

File: lighthouse-core/report/html/renderer/report-renderer.js [hunk 6]
```diff
@@ -136,7 +130,7 @@ class ReportRenderer {
 
     const container = this._dom.cloneTemplate('#tmpl-lh-warnings--toplevel', this._templateContext);
     const message = this._dom.find('.lh-warnings__msg', container);
-    message.textContent = Util.UIStrings.toplevelWarningsMessage;
+    message.textContent = report.i18n.rendererFormattedStrings.toplevelWarningsMessage;
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/report-renderer.js]: forgot to update?

File: lighthouse-core/report/html/renderer/report-ui-features.js [hunk 7]
```diff
@@ -252,7 +252,7 @@ class ReportUIFeatures {
       this._dom.find('.lh-3p-filter-count', filterTemplate).textContent =
           `${thirdPartyRows.size}`;
       this._dom.find('.lh-3p-ui-string', filterTemplate).textContent =
-          Util.UIStrings.thirdPartyResourcesLabel;
+          this.json.i18n.rendererFormattedStrings.thirdPartyResourcesLabel;
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/report-ui-features.js]: forgot to update?

File: lighthouse-core/report/html/renderer/i18n.js [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    });
+
+    return parts.join(' ');
+  }
+
+  /**
+   * Set the locale to be used for I18n's number and date formatting functions.
+   * @param {LH.Locale} locale
+   */
+  setNumberDateLocale(locale) {
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/i18n.js]: seems like we can get rid of this and just keep the `locale` fixing part in the constructor?

File: lighthouse-core/audits/dobetterweb/dom-size.js [hunk 9]
```diff
@@ -97,28 +96,30 @@ class DOMSize extends Audit {
       {key: 'value', itemType: 'numeric', text: str_(UIStrings.columnValue)},
     ];
 
+    const i18n = new I18n(context.settings.locale);
```
Comment (patrickhulce) [inline lighthouse-core/audits/dobetterweb/dom-size.js]: it's a bummer these won't get into the localeLog, maybe we file a separate issue to track that?

Comment (connorjclark) [reply lighthouse-core/audits/dobetterweb/dom-size.js]: no idea what this means :)

Comment (patrickhulce) [reply lighthouse-core/audits/dobetterweb/dom-size.js]: Oh sorry I meant [icuMessagePaths](https://github.com/GoogleChrome/lighthouse/blob/39b13bc4f48cf6933f28f3ade87e1705f18a2ddb/lighthouse-core/runner.js#L152) 

when we create i18n'd strings using `str` we get to track it in the lhr so we can use `swap-locale` later, it'd be great if these usages could take advantage of that

Comment (connorjclark) [reply lighthouse-core/audits/dobetterweb/dom-size.js]: ah right. an additional challenge - besides figuring out how to sink these paths to a central place - the swap locale script only works for UIStrings, not for value formatting.

Instead, I think we shouldn't format in the audit at all. These values should be raw numbers - and all the `numeric` detail types should be formatted by the report renderer. That way there's nothing to swap.

Comment (connorjclark) [reply lighthouse-core/audits/dobetterweb/dom-size.js]: does that make sense?

Comment (patrickhulce) [reply lighthouse-core/audits/dobetterweb/dom-size.js]: oh sorry! yes! 

> I think we shouldn't format in the audit at all. These values should be raw numbers - and all the numeric detail types should be formatted by the report renderer (odd that this isn't the case now ...). That way there's nothing to swap.

Agreed I think there was even a todo I put in one of my PRs during review on this topic that I've totally forgotten to merge...

Comment (connorjclark) [reply lighthouse-core/audits/dobetterweb/dom-size.js]: Cool.. so I'll leave the changes in the audits as they are now. It's still an  improvement. We can follow up on formatting in just the renderer later.

I'll fix those tests now..

File: lighthouse-core/report/html/renderer/i18n.js [hunk 10]
```diff
@@ -0,0 +1,142 @@
+/**
+ * @license
+ * Copyright 2020 Google Inc. All Rights Reserved.
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/i18n.js]: :tada:

**Comment by connorjclark** [comment]
> This is a good incremental improvement, but is it part of a larger change? I feel like it is and I'm missing the context from the PR desc.

yea https://github.com/GoogleChrome/lighthouse/pull/9166/files#r361547882

File: lighthouse-core/report/html/renderer/i18n.js [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  get strings() {
+    return this._strings;
+  }
+
+  /**
+   * Format number.
+   * @param {number} number
+   * @param {number=} granularity Number of decimal places to include. Defaults to 0.1.
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/i18n.js]: don't want to change functionality in this pr

[the remark this answers is not in the document]
Comment (exterkamp) [reply lighthouse-core/report/html/renderer/i18n.js]: None of the unit-tests use granularity except 1. I'm fine if it's a follow up/not done, just seems like old YAGNI.

[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/i18n.js]: ah, i assumed.

`detailsRenderer` uses this a bit.

File: lighthouse-core/report/html/renderer/category-renderer.js [hunk 12]
```diff
@@ -115,10 +116,11 @@ class CategoryRenderer {
     if (audit.result.scoreDisplayMode === 'error') {
       auditEl.classList.add(`lh-audit--error`);
       const textEl = this.dom.find('.lh-audit__display-text', auditEl);
-      textEl.textContent = Util.UIStrings.errorLabel;
+      textEl.textContent = strings.errorLabel;
       textEl.classList.add('tooltip-boundary');
       const tooltip = this.dom.createChildOf(textEl, 'div', 'tooltip tooltip--error');
-      tooltip.textContent = audit.result.errorMessage || Util.UIStrings.errorMissingAuditInfo;
+      tooltip.textContent =
+        audit.result.errorMessage || strings.errorMissingAuditInfo;
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/category-renderer.js]: didn't this line get shorter? must be a leftover newline from earlier version :)

File: lighthouse-core/report/html/renderer/i18n.js (hunk 10 above)
Comment (patrickhulce) [reply lighthouse-core/report/html/renderer/i18n.js]: you're avoiding the condensed version though?? 😢 😢

File: lighthouse-core/report/html/renderer/report-renderer.js [hunk 13]
```diff
@@ -185,6 +179,13 @@ class ReportRenderer {
    * @return {DocumentFragment}
    */
   _renderReport(report) {
+    const i18n = new I18n(report.configSettings.locale, {
+      // Set missing renderer strings to default (english) values.
+      ...Util.UIStrings,
+      ...report.i18n.rendererFormattedStrings,
+    });
+    Util.i18n = i18n;
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/report-renderer.js]: I'm a little concerned that somewhere, someday we're going to do the `const strings = Util.i18n.strings` shortcut in a place that doesn't happen on every render and we're going to have a subtle bug we'll probably never catch, but I don't have *great* suggestions around this, `.strings` returns an object that proxies properties to the latest Util.i18n._strings prop? 

any thoughts on this?

File: lighthouse-core/report/html/renderer/i18n.js (hunk 10 above)
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/i18n.js]: lol didnt notice, i just here to copy and paste stuff man

File: lighthouse-core/report/html/renderer/report-renderer.js (hunk 13 above)
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/report-renderer.js]: that would require stashing that value in what is effectively a global (if it isn't being reset every render i assume you're talking about a global). I think good PRs will prevent that.

Comment (patrickhulce) [reply lighthouse-core/report/html/renderer/report-renderer.js]: > if it isn't being reset every render i assume it must be a global

This assumption during reviewing is another reason I worry about this :)
